compile-time-programming/requires-expressions.md

```diff
@@ -0,0 +1,102 @@
+# Module name: Requires Expressions
+
+## Overview
+
+<table>
+  <thead>
+    <th>Level</th>
+    <th>Objectives</th>
+  </thead>
+  <tr>
+    <td>Foundational knowledge</td>
+    <td>Define and use requires-expressions to check satisfaction of expressions by given parameters</td>
+  </tr>
+  <tr>
+    <td>Advanced</td>
+    <td>Define and use requires-expressions to check properties of expressions</td>
+  </tr>
+  <tr>
+    <td>Further studies</td>
+    <td></td>
+  </tr>
+</table>
+
+## Motivation
+
+Requires-expressions allow a developer to perform compile-time evaluation on the validity of other expressions. These are fundamental to the ability to write concepts.[[Compile-time programming: concepts]][1]
+
+## Topic introduction
+
+Requires-expressions are compile-time predicates which evaluate to true when their specified set of expressions are all valid for a given set of inputs.
+
+## Foundational knowledge: Writing requires-expressions
+
+### Background/Required Knowledge
+
+A student is able to:
+
+* Write and use a function template [[Compile-time programming: function templates]][2]
+* Differentiate between expressions and statements
+
+It is helpful if:
+
+* A student is aware of the constraints imposed by their template definitions
+
+### Student outcomes
+
+A student should be able to:
+
+1. Write a simple-requirement to assert the validity of an expression
+2. Write a type-requirement to check the existence of a type by its identifier
+3. Write a compound-requirement to test the resulting type of an expression
+4. Write a nested-requirement to test the constexpr value of an operation, as opposed to just the semantic validity
+5. Use a requires-expression within a concept, requires-clause, or `if constexpr` condition
```

Comment:
I'm not convinced these are foundational.

Reply:
My intent was to specify the students ability to utilize the requires-expression in _some_ way. These were the only three places I could think of which would supply a motivating usage of a requires expression. Thus, I thought to mention that using the requires expression within at least _one_ of these would be a foundational outcome.

Reply:
The names such as "compound-requirement" and "nested-requirement" are certainly not foundational, or even needed IMO. Certainly the _instructor_ is not likely to know what these jargon terms mean.  Is it possible to rephrase these without the jargon, e.g.

    1. Write a simple requires-expression that tests whether `t + 1` is well-formed
    2. Write a requires-expression that tests whether the nested type `T::value_type` exists
    3. Write a requires-expression that tests whether the type of `t + 1` is exactly `int`;
        whether it's a scalar type
    4. Write a requires-expression that tests whether `T::value` is `true`
        (not just whether it's well-formed)
    5. Use a requires-expression as a boolean expression within a concept definition,
        requires-clause, or condition

I dunno, am I just trading your preferred jargon ("compound-requirement") for mine ("well-formed")?

Reply:
I presume our goal is not to _teach_ the instructors, so much as to guide them about what should be taught? I would wholly expect an instructor to need to do some research on the topic in order to teach it, separate from these guides. Through that, shouldnt we use the proper terms so that they can actually stand a chance in learning enough to teach?